ckanext/dcat/validators.py

```diff
@@ -0,0 +1,134 @@
+import datetime
+import json
+import re
+
+from ckantoolkit import (
+    missing,
+    StopOnError,
+    get_validator,
+    Invalid,
+    _,
+)
+from ckanext.scheming.validation import scheming_validator
+
+# https://www.w3.org/TR/xmlschema11-2/#gYear
+regexp_xsd_year = re.compile(
+    "-?([1-9][0-9]{3,}|0[0-9]{3})(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))?"
+)
+
+# https://www.w3.org/TR/xmlschema11-2/#gYearMonth
+regexp_xsd_year_month = re.compile(
+    "-?([1-9][0-9]{3,}|0[0-9]{3})-(0[1-9]|1[0-2])(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))?"
+)
+
+regexp_xsd_date = re.compile(
+    "-?([1-9][0-9]{3,}|0[0-9]{3})-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))?"
+)
+
+
+def is_year(value):
+    return regexp_xsd_year.fullmatch(value)
+
+
+def is_year_month(value):
+    return regexp_xsd_year_month.fullmatch(value)
+
+
+def is_date(value):
+    return regexp_xsd_date.fullmatch(value)
+
+
+def dcat_date(key, data, errors, context):
+    value = data[key]
+
+    scheming_isodatetime = get_validator("scheming_isodatetime")
+
+    if isinstance(value, datetime.datetime):
+        return
+
+    if is_year(value) or is_year_month(value) or is_date(value):
+        return
+
+    try:
+        scheming_isodatetime({}, {})(key, data, errors, context)
+    except Invalid:
+        raise Invalid(
+            _(
+                "Date format incorrect. Supported formats are YYYY, YYYY-MM, YYYY-MM-DD and YYYY-MM-DDTHH:MM:SS"
```

Comment:
This is overly restrictive. The first datetime I attempted to parse with the dcat harvester was of the form: `YYY-MM-DDTHH:MM:SS.000Z`, which is permitted by ISO8601 (https://en.wikipedia.org/wiki/ISO_8601) and the `xsd:dateTime` (https://www.w3.org/TR/xmlschema11-2/#dateTime) spec. 

It appears that this goes back to the ckan helper  date_str_to_datetime which "Converts an ISO _like_ date to a datetime", using a 12 yr old regex based time zone ignoring date parser. This predates python having reasonable sensible timezone aware datetimes and the `datetime.datetime.fromisoformat` function. 

The core helper probably should be fixed, with its potential for knock on changes, but in the meantime since this is new code, perhaps we should just directly use `datetime.datetime.fromisoformat` here. (python3.7 min for that). We'd still need to support `YYYY` and `YYYY-MM` specially in the code, because those aren't covered

Reply:
[`datetime.datetime.fromisoformat()`](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat) is quite limited up until Python 3.11:

> Changed in version 3.11: Previously, this method only supported formats that could be emitted by [date.isoformat()](https://docs.python.org/3/library/datetime.html#datetime.date.isoformat) or [datetime.isoformat()](https://docs.python.org/3/library/datetime.html#datetime.datetime.isoformat).

So on python 3.10 and lower these dates are not parsed:

```
Python 3.10.10 (main, Mar 14 2023, 15:55:23) [GCC 9.4.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.13.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import datetime

In [2]: datetime.datetime.fromisoformat("2011-11-04T00:05:23Z")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[2], line 1
----> 1 datetime.datetime.fromisoformat("2011-11-04T00:05:23Z")

ValueError: Invalid isoformat string: '2011-11-04T00:05:23Z'
```

What do you think of the approach I followed in https://github.com/ckan/ckanext-dcat/pull/281/commits/c7b8c027495f4dece84562090b44fd20f1374604? If something is not an `xsd:gYear`, an `xsd:gYearMonth` or an `xsd:date` we just let dateutil parse it (which will accept the timezone values you suggested). If that parses we serve it as an `xsd:dateTime`.
The tests check that 1) the input value is unchanged and 2) it's served with the correct `xsd` data type.

Reply:
Looks better than doing it ourselves. You'r enot testing for invallid dates, or ... conditionally valid dates like M/D/Y, but overall, I'm definitely happier with getting a builtin date parser that handles the in the wild formats that I've seen.

Reply:
Added tests in [39b4d91](https://github.com/ckan/ckanext-dcat/pull/281/commits/39b4d9162c30772b4c10db17ee7126b734c95319)